src/main/scala/vectorpipe/osm/internal/ProcessOSM.scala

```diff
@@ -0,0 +1,769 @@
+      .withColumn("geometryChanged", !((lag('geom, 1) over idByUpdated) <=> 'geom))
+      .join(ways.select('id, 'version, 'tags, 'visible, 'uid, 'user), Seq("id", "version"))
+      .select(
+        lit(WayType) as '_type,
+        'id,
+        'geom,
+        'tags,
+        'changeset,
+        'updated,
+        'validUntil,
+        'uid,
+        'user,
+        'visible,
+        'version,
+        'minorVersion,
+        'geometryChanged)
+  }
+
+  private def getRelationMembers(relations: DataFrame, geoms: DataFrame) = {
+    implicit val ss: SparkSession = relations.sparkSession
+    import ss.implicits._
+
+    // way to relation lookup table (missing 'role, since it's not needed here)
+    val waysToRelations = relations
+      .select(explode('members) as 'member, 'id as 'relationId, 'version, 'timestamp, 'validUntil)
+      .withColumn("type", $"member.type")
+      .withColumn("id", $"member.ref")
+      .drop('member)
+
+    @transient val idByVersion = Window.partitionBy('id).orderBy('version)
+
+    // Create a relation entry for each changeset in which a geometry was modified, containing the timestamp and
+    // changeset of the geometry that triggered the association. This will later be used to assemble relations at each
+    // of those points in time.
+    // If you need authorship, join on changesets
+    val relationsByChangeset = geoms
+      .where('geometryChanged)
+      .drop('validUntil)
+      // re-calculate validUntil windows
+      .withColumn("validUntil", lead('updated, 1) over idByVersion)
+      // TODO when expanding beyond relations referring to ways, geoms should include 'type for the join to work properly
+      .withColumn("type", lit(WayType))
+      .select('type, 'changeset, 'id, 'updated, 'uid, 'user)
+      .join(waysToRelations, Seq("id", "type"))
+      .where(waysToRelations("timestamp") <= geoms("updated") and
+        geoms("updated") < coalesce(waysToRelations("validUntil"), current_timestamp))
+      .select('changeset, 'relationId as 'id, 'version, 'updated, 'uid, 'user)
+
+    @transient val idAndVersionByUpdated = Window.partitionBy('id, 'version).orderBy('updated)
+    @transient val idByUpdated = Window.partitionBy('id).orderBy('updated)
+
+    val allRelationVersions = relationsByChangeset
+      // Union with raw relations to include those in the time line (if they weren't already triggered by geometry
+      // modifications at the same time)
+      .union(relations.select('changeset, 'id, 'version, 'timestamp as 'updated, 'uid, 'user))
+      // If a node, a way, and/or a relation were modified within the same changeset at different times, there will be
+      // multiple entries with different timestamps; this reduces them down to a single update per changeset.
+      .groupBy('changeset, 'id)
+      .agg(max('version).cast(IntegerType) as 'version, max('updated) as 'updated)
+      .join(relations.select('id, 'version, 'members, 'uid, 'user), Seq("id", "version"))
```

Comment:
This has the same issue as ways where authorship will be incorrect for minor versions.

There's another, not obvious at all (but alluded to in the comment below), problem that surfaced when generating geometries for the full planet history: in order to calculate `validUntil`, **all** relation rows need to be on the same partition for the window function to execute.

tl;dr: this function is very sensitive to the addition of new columns, especially columns that can be large, so it's worth seeing if data can be added after the explosion / construction stage.

Reply:
Oh, I see. I didn't realize that there was a performance cost. I'll remove those extra columns then. Would adding the `uid` and `user` information after the dataframe is created be efficient? I was thinking about joining them by using the `id`s from the data source.

Reply:
Yes. Using `addUserMetadata` on the output of this function should do the trick.